### PR TITLE
Switch building airlfow packages to generic images instead of CI image

### PR DESCRIPTION
**Pull request by potiuk**
Since we are using setuptools (for now) to build airflow packages,
we still want to run the build in docker image in order to not take
the risk that code from forks is executed as part of the job that can
have access to secrets and package writing.

However we do not need the whole Airflow CI image for that. We can
use two different images:

* official Node image to compile assets on CI
* Custom docker image based on Python, where we only install git,
  gitpython and wheels/pip in the right versions in order to retrieve
  commit hash

Since compiling assets via mounted volume on MacOS is slow, we optimize
the build and assets on MacOS are compiled locally using pre-commits -
but in CI / on Linux we compile assets using the Node image.

This PR switches the images used and uses Python script to build the
packages rather than bash script.

Building PROD image is also faster now in CI as it does not need to pull
CI image to build packages - so we can remove the step of pulling the
CI image needed for PROD build. We still need to wait for the
constraints to be prepared so PROD build waits for CI image to complete,
but it does not pull nor use the image.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: scripts/in_container/_in_container_utils.sh [hunk 1]
```diff
@@ -105,29 +105,6 @@ function in_container_basic_sanity_check() {
     in_container_go_to_airflow_sources
 }
 
-export DISABLE_CHECKS_FOR_TESTS="missing-docstring,no-self-use,too-many-public-methods,protected-access,do-not-use-asserts"
```
Comment (potiuk) [inline scripts/in_container/_in_container_utils.sh:108]: Removed unused code.

**Comment by potiuk** [comment]
cc: @ephraimbuddy - I would love that one to be merged before 2.8.0 cut-off as it finally closes some of the potential security loops we had in our release process and potential ways how our package could be potentially modified at release time - would be great if you can also see if it works nicely for you when you generate packages in release process - it changes slightly the way where airflow packages are generated -  previously they were generated:

* static assets were generated in local pre-commit prepared environment
* the airflow packages were generated inside Breeze CI image

This change implements a bit different execution environment for release preparation: 

* static assets were generated in local pre-commit prepared environment (for performance reason) but on Linux (and I think I will also add flag for release manager to run it also on MacOS even if it is slower) - in a separate official node image

* the airflow packages is built in a separate Docker container that is much smaller and faster and always locally built without using Breeze CI image (It is just Python official image + installed git + wheel + pip + setuptools + rich for diagnostics)

The change allows to avoid some (rather complex and difficult to pull of) scenarios where (breaking some other things) malicious users could influence the content of prepared packages. 

So it would be great to get that in for 2.8.0 at the moment of branch cut-off

**Comment by ephraimbuddy** [comment]
> cc: @ephraimbuddy - I would love that one to be merged before 2.8.0 cut-off as it finally closes some of the potential security loops we had in our release process and potential ways how our package could be potentially modified at release time - would be great if you can also see if it works nicely for you when you generate packages in release process - it changes slightly the way where airflow packages are generated - previously they were generated:
> 
> * static assets were generated in local pre-commit prepared environment
> * the airflow packages were generated inside Breeze CI image
> 
> This change implements a bit different execution environment for release preparation:
> 
> * static assets were generated in local pre-commit prepared environment (for performance reason) but on Linux (and I think I will also add flag for release manager to run it also on MacOS even if it is slower) - in a separate official node image
> * the airflow packages is built in a separate Docker container that is much smaller and faster and always locally built without using Breeze CI image (It is just Python official image + installed git + wheel + pip + setuptools + rich for diagnostics)
> 
> The change allows to avoid some (rather complex and difficult to pull of) scenarios where (breaking some other things) malicious users could influence the content of prepared packages.
> 
> So it would be great to get that in for 2.8.0 at the moment of branch cut-off

Sure. Whatever is merged from now till Monday will be part of the beta release. I will continuously rebase the test branch till Monday

**Comment by potiuk** [comment]
I've added the `--use-container-for-assets-compilation` flag (and added it also to the release process).

**Comment by potiuk** [comment]
Also as discussed with @ephraimbuddy on slack -> this one will make life easier for release manager that has weak/unstable/not the fastest network.

It uses WAY smaller images (5x) than before - with only node (slim) and python (slim) - both based on bookworm debian - with the minium set of dependencies needed to build the package. 

That has the nice effect of having a reproducible "package build" environment that is based on only official images and packages but also it should not change as often as CI breeze image.

Fixing all dependencies and versions bring us closer to have reproducible builds for airflow package - it is not yet there and it will likely require changing the build tooling from setup tools to hatch or flit, and solving reproducibility of  asset compilation, but at least with this change we have really fast way to rebuild Airflow package locally in a fully repeatable way without the dangers coming from using caching when building Breeze CI image.

**Review by ephraimbuddy — APPROVED** [review]

**Pull request merged**